### ISSUE TITLE
Tracking issue for StreamObservers

### ISSUE THREAD

**Issue by carl-mastrangelo** [grpc/grpc-java#4694]

**Comment by micheljung** [comment]
Has anyone ever used `StreamObservers` successfully? I doubt that.

The JavaDoc says:

> on servers it is safe to call inside the service method implementation.

So you implement something like this (Kotlin):
```
  override fun get(
    request: DummyRequest,
    responseObserver: StreamObserver<DummyResponse>,
  ) {
    val counter = AtomicInteger()
    val iterator = Stream.generate { counter.getAndIncrement() }
      .map { dummyResponse { value = it } }
      .iterator()

    StreamObservers.copyWithFlowControl(
      iterator,
      responseObserver as CallStreamObserver<DummyResponse>,
    )
  }
```

but by the time `get()` is called, the stream is already `ready`. Therefore, the `onReadyHandler` set by `copyWithFlowControl`:

    target.setOnReadyHandler(new FlowControllingOnReadyHandler());

is never called because `setOnReadyHandler` doesn't trigger the handler and there is no state change to `ready`.

**Comment by micheljung** [comment]
If you have a method like this:

So you implement something like this (Kotlin):
```kotlin

  @Transactional
  override fun get(
    request: DummyRequest,
    responseObserver: StreamObserver<DummyResponse>,
  ) {
    val counter = AtomicInteger()
    myStreamingDatabaseQuery()
      .iterator()
      .forEach { responseObserver.onNext(it) }

    responseObserver.onCompleted()
  }

  fun myStreamingDatabaseQuery(): Stream<Dummy> = ...
```

It will work fine as the transaction is opened, the stream is processed, and the transaction is closed.

However, you might then run into memory issues and try to use `StreamObservers` like this:

```kotlin
  @Transactional
  override fun get(
    request: DummyRequest,
    responseObserver: StreamObserver<DummyResponse>,
  ) {
    val counter = AtomicInteger()
    val iterator = myStreamingDatabaseQuery().iterator()

    StreamObservers.copyWithFlowControl(
      iterator,
      responseObserver as CallStreamObserver<DummyResponse>,
    )
  }

  fun myStreamingDatabaseQuery(): Stream<Dummy> = ...
```

This won't work because `StreamObservers` starts processing the stream **after** `get()` has returned. However, the transaction will then have been closed and you can no longer read the stream.

Any idea how to solve this?

**Comment by micheljung** [comment]
The issue extends to closing the stream, which my code above lacks.

**Comment by ejona86** [comment]
`@Transactional` assumes all processing is done within the method. That requires the method to be blocking. But for streaming you have to implement the service method in an async fashion (in order to get onReady() callbacks). So `@Transactional` is simply incompatible with gRPC streaming until we have some blocking API on server-side. But that's not specific to the `StreamObservers` utility.

**Comment by larry-safran** [comment]
Agreed to mark it deprecated because it is of minimal usefulness and is not frequently used (nothing in g3 or sourcegraph).   Prefer to have grpc-java usages of it replaced, instead of creating an iterator just use onReady.

**Comment by ejona86** [comment]
We should close this when the class is actually removed, which is also the point the reference to this issue is removed.
